pkg/operator/operator.go

```diff
@@ -241,76 +273,213 @@ func (k *Kontroller) withLeaderElection() error {
 func (k *Kontroller) process() {
 	glog.V(4).Info("Going through a loop cycle")
 
-	nodelist, err := k.nc.List(v1meta.ListOptions{})
+	// find nodes with the after-reboot=true label and check if all provided
+	// annotations are set. if all annotations are set to true then remove the
+	// after-reboot=true label and set reboot-ok=false, telling the agent that
+	// the reboot has completed.
+	err := k.checkAfterReboot()
 	if err != nil {
-		glog.Infof("Failed listing nodes %v", err)
+		glog.Errorf("Failed to check after reboot: %v", err)
 		return
 	}
 
-	glog.V(6).Infof("Found nodes: %+v", nodelist.Items)
+	// find nodes which just rebooted but haven't run after-reboot checks.
+	// remove any old annotations and add the after-reboot=true label.
+	err = k.markAfterReboot()
+	if err != nil {
+		glog.Errorf("Failed to update recently rebooted nodes: %v", err)
+		return
+	}
 
-	justRebootedNodes := k8sutil.FilterNodesByAnnotation(nodelist.Items, justRebootedSelector)
+	// find nodes with the before-reboot=true label and check if all provided
+	// annotations are set. if all annotations are set to true then remove the
+	// before-reboot=true label and set reboot=ok=true, telling the agent it's
+	// time to reboot.
+	err = k.checkBeforeReboot()
```

Comment:
I like the idea of putting them in goroutines to enforce the fact that the ordering doesn't matter, but I feel like that is asking for weird race condition bugs. The ordering of the state checking and modification doesn't matter, but it would take a lot more work to make us confident that the functions are safe to run concurrently. 

I agree with the error returning though. I'll change that right now.